Fix issues

Keyrxng opened the pull request:
Resolves #21

Below is the activity logged after `lastCheck`.


old results from [110](https://github.com/ubiquibot/conversation-rewards/issues/110)

```yml
[
  {
    event: 'review_requested',
    created_at: '2024-09-15T00:21:13Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'review_requested',
    created_at: '2024-09-14T05:58:45Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'review_requested',
    created_at: '2024-09-14T05:58:44Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'review_requested',
    created_at: '2024-09-14T05:58:44Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'ready_for_review',
    created_at: '2024-09-14T05:57:21Z',
    actor: 'gentlementlegen'
  }
]
``` 

new results:

```yml
[
  {
    event: 'review_requested',
    created_at: '2024-09-15T00:21:13Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'committed',
    created_at: '2024-09-15T00:16:33Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'review_requested',
    created_at: '2024-09-14T05:58:45Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'review_requested',
    created_at: '2024-09-14T05:58:44Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'review_requested',
    created_at: '2024-09-14T05:58:44Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'ready_for_review',
    created_at: '2024-09-14T05:57:21Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'committed',
    created_at: '2024-09-14T05:55:14Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'committed',
    created_at: '2024-09-14T05:50:43Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'committed',
    created_at: '2024-09-14T05:20:33Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'committed',
    created_at: '2024-09-13T20:57:37Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'committed',
    created_at: '2024-09-13T20:48:44Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'committed',
    created_at: '2024-09-13T19:04:53Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'committed',
    created_at: '2024-09-13T16:35:06Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'committed',
    created_at: '2024-09-13T16:12:54Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'committed',
    created_at: '2024-09-13T16:11:02Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'committed',
    created_at: '2024-09-13T16:06:00Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'committed',
    created_at: '2024-09-13T16:00:00Z',
    actor: 'gentlementlegen'
  },
  {
    event: 'committed',
    created_at: '2024-09-13T15:53:56Z',
    actor: 'gentlementlegen'
  }
]
```

0x4007 commented on src/helpers/get-assignee-activity.ts:
```diff
@@ -27,7 +27,50 @@ export async function getAssigneesActivityForIssue(context: Context, issue: List
+  for (const event of issueEvents) {
+    let actorId = null;
+    let actorLogin = null;
+    let createdAt = "UNKNOWN";
```
Unknown seems like an unusual value. Should probably be undefined.

0x4007 commented on src/helpers/task-deadline.ts:
```diff
@@ -35,13 +38,17 @@ export async function getDeadlineWithThreshold(
-  let deadlineWithThreshold = deadline.plus({ milliseconds: config.disqualification });
-  let reminderWithThreshold = deadline.plus({ milliseconds: config.warning });
+  const filteredActivity = activity.filter((o) => {
+    return eventWhitelist.includes(o.event || "");
```
Empty string seems suspicious

gentlementlegen commented on src/types/plugin-inputs.ts:
```diff
@@ -49,6 +56,10 @@ export const userActivityWatcherSettingsSchema = T.Object({
+   * List of events to consider as valid activity on a task
+   */
+  eventWhitelist: T.Array(T.String({ default: Object.values(eventWhitelist) })),
 });
```
```suggestion
}, { default: {} }});
```
You will need this otherwise an empty configuration will fail.

Keyrxng replied:
My mistake ty

gentlementlegen reviewed (commented):
Could you please update README.md since I think the configuration changed?

gentlementlegen commented on src/types/plugin-inputs.ts:
```diff
@@ -31,24 +31,38 @@ function thresholdType(options?: StringOptions) {
+    /**
+     * List of events to consider as valid activity on a task
+     */
+    eventWhitelist: T.Array(T.String({ default: Object.values(eventWhitelist) })),
```
```suggestion
    eventWhitelist: T.Array(T.String(), { default: Object.values(eventWhitelist) }),
```

Otherwise if `eventWhitelist` is missing the configuration validation will fail. Could you add a test for a default configuration?

Keyrxng commented:
@gentlementlegen tests added and issue resolved

gentlementlegen commented:
The configuration is fixed, but I am a bit lost with the current behavior of that plugin after testing it. I do not know if that is introduced by the changes of this pull-request, but look at [this run](https://github.com/Meniole/user-activity-watcher/actions/runs/10943790050/job/30384103248#step:5:31):
- I just started a task by assigning myself
- the deadline according to the logs is in the past
- I get unassigned right away

Keyrxng commented:
So because you used the /start command last month, then just now assigned via UI self-assign, it has taken the deadline from the what I assume is the comment as opposed to your self-assignment event.

```json
{
  "now": "Sep 19, 2024, 3:09 PM",
  "reminder": "Aug 22, 2024, 9:37 AM",
  "deadline": "Aug 22, 2024, 9:38 AM",
  "caller": "updateTaskReminder"
}
```

So what's the solution? Needs a little debugging to see what the best way to handle it would be, something like use the timeline first and fallback to the comments. UI assignment means the actor is the user, start or direct PR would be the bot. I think best handled in another task though.

Keyrxng commented:
Likely myself who introduced it @gentlementlegen. Realized sort was backwards, but your example highlighted the fact we should use the timeline over comments if the timeline is later than the lastCheck so I implemented that and updated tests

![image](https://github.com/user-attachments/assets/930888d1-786c-43c7-8cab-85c6c283272f)

gentlementlegen commented:
@Keyrxng I pulled your latest changes and ran again, the behavior was similar
> Do we as an org permit contributors to self-assign? It bypasses checks in other plugins.

This applies only to members but we could potentially be un-assigned for no reason the way it is happening to me right now. Can't we only based the checks to use the timeline events and check the last time a user was assigned?

0x4007 commented:
Ignore the deadline. 

Unassignment only should happen after 7 days of no activity (configurable)

0x4007 commented on README.md:
```diff
@@ -43,4 +43,9 @@ yarn test
       optOut:
         - "repoName"
         - "repoName2"
+    eventWhitelist: # these are the tail of the webhook event i.e pull_request.review_requested
```
Should just use the full identifier. Shortening just will lead to confusion.

Keyrxng replied:
So the config would take the full `pull_request.review_requested` and I'll handle string splitting, is that what you mean?

0x4007 replied:
It doesn't look like this was ever updated in the readme.

0x4007 replied:
Looks like this was never fixed. Should use full webhook names.

https://github.com/ubiquity-os-marketplace/daemon-disqualifier/blob/9ca3f7ba20a48385151a55e7cf52e57cafa9cf65/tests/main.test.ts#L67

Keyrxng commented:
QA: https://github.com/ubq-testing/user-activity-watcher/actions/runs/11016118749/job/30590832730

0x4007 commented on src/helpers/task-update.ts:
```diff
@@ -3,42 +3,43 @@ import { Context } from "../types/context";
   }
 
   if (now >= deadlineWithThreshold) {
+    // if the issue is past due, we should unassign the user
```
Ignore deadline. This is strictly unassigning based on config idle timeout

Keyrxng replied:
`deadlineWithThreshold` here is either `lastActivity.createdAt + threshold` or `(issue.createdAt + labelDuration) + threshold` or `issue.created + threshold`. So this is unassigning based on the configurable idle timeout setting

0x4007 replied:
You should remove misleading comments.

0x4007 reviewed (changes requested):
It's not clear to me if you're ignoring the deadline as I requested, and if you are actually checking the assign event on the timeline, instead of checking the metadata from the /start response.

DeFi-crypto commented:
This looks really interesting. I'm trying to try the plugin currently and it seems to be working fine.

0x4007 commented on src/types/plugin-inputs.ts:
```diff
@@ -31,24 +31,94 @@ function thresholdType(options?: StringOptions) {
+      .Encode((value) =>
+        value.map((event) => {
+          const roleMap: Map<TimelineEvents, WhitelistEvents> = new Map([
+            ["review_requested", "pull_request.review_requested"],
+            ["ready_for_review", "pull_request.ready_for_review"],
+            ["commented", "pull_request_review_comment.created"],
+            ["commented", "issue_comment.created"],
+            ["committed", "push"],
```
Are these defaults?

0x4007 commented on src/helpers/task-metadata.ts:
```diff
@@ -77,9 +58,13 @@ export async function getTaskMetadata(
+  metadata.startPlusLabelDuration =
+    DateTime.fromISO(mostRecentAssignmentEvent?.created_at || issue.created_at)
+      .plus({ milliseconds: durationInMs })
+      .toISO() || "";
```
Setting the duration to an empty string looks wrong.

0x4007 reviewed (approved):
Can address remaining problems later. The plugin is quite dysfunctional in production and I think this will get most of the essential functionality back up.

Unfortunately due to merge conflicts this won't be able to be merged.
